### PR TITLE
Invert null buffer setting

### PR DESCRIPTION
**Pull request by nevi-me**
Implements inverting null buffers for numeric and boolean arrays, yields more improvement for the arrays.

```rust
Gnuplot not found, using plotters backend
take i32 512            time:   [1.3542 us 1.3586 us 1.3638 us]
                        change: [-66.653% -66.483% -66.322%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

take i32 1024           time:   [2.2913 us 2.2976 us 2.3043 us]
                        change: [-70.896% -70.739% -70.592%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

take bool 512           time:   [1.3893 us 1.4114 us 1.4374 us]
                        change: [-56.775% -56.346% -55.805%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe

take bool 1024          time:   [2.1283 us 2.1342 us 2.1401 us]
                        change: [-68.305% -68.011% -67.738%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

take str 512            time:   [2.6701 us 2.6780 us 2.6877 us]
                        change: [-77.293% -77.103% -76.907%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  9 (9.00%) high mild
  2 (2.00%) high severe

take str 1024           time:   [7.0615 us 7.0996 us 7.1548 us]
                        change: [-65.804% -65.687% -65.526%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
```

### CODE REVIEW

File: rust/arrow/src/compute/kernels/take.rs [hunk 1]
```diff
@@ -171,15 +173,15 @@ where
     let array = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
 
     let num_bytes = bit_util::ceil(data_len, 8);
-    let mut null_buf = MutableBuffer::new(num_bytes).with_bitset(num_bytes, false);
+    let mut null_buf = MutableBuffer::new(num_bytes).with_bitset(num_bytes, true);
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/take.rs:176]: don't know why the diff includes your changes too, I'll manually mark the changes that I made. 1

File: rust/arrow/src/compute/kernels/take.rs [hunk 2]
```diff
@@ -171,15 +173,15 @@ where
...
 
     let null_slice = null_buf.data_mut();
 
     let new_values: Vec<T::Native> = (0..data_len)
         .map(|i| {
             let index = indices.value(i) as usize;
-            if array.is_valid(index) {
-                bit_util::set_bit(null_slice, i);
+            if !array.is_valid(index) {
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/take.rs:183]: 2

File: rust/arrow/src/compute/kernels/take.rs [hunk 3]
```diff
@@ -209,7 +211,7 @@ fn take_boolean(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef> {
     let array = values.as_any().downcast_ref::<BooleanArray>().unwrap();
 
     let num_byte = bit_util::ceil(data_len, 8);
-    let mut null_buf = MutableBuffer::new(num_byte).with_bitset(num_byte, false);
+    let mut null_buf = MutableBuffer::new(num_byte).with_bitset(num_byte, true);
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/take.rs:214]: 3

File: rust/arrow/src/compute/kernels/take.rs [hunk 4]
```diff
@@ -218,10 +220,11 @@ fn take_boolean(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef> {
     (0..data_len).for_each(|i| {
         let index = indices.value(i) as usize;
         if array.is_valid(index) {
-            bit_util::set_bit(null_slice, i);
             if array.value(index) {
                 bit_util::set_bit(val_slice, i);
             }
+        } else {
+            bit_util::unset_bit(null_slice, i);
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/take.rs:227]: 4

File: rust/arrow/src/compute/kernels/take.rs [hunk 5]
```diff
@@ -297,29 +307,26 @@ fn take_list(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef> {
     let list: &ListArray = values.as_any().downcast_ref::<ListArray>().unwrap();
     let (list_indices, offsets) = take_value_indices_from_list(values, indices);
     let taken = take(&list.values(), &list_indices, None)?;
-    // determine null count and null buffer, which are a function of `values` and `indices`
-    let mut null_count = 0;
-    let num_bytes = bit_util::ceil(indices.len(), 8);
-    let mut null_buf = MutableBuffer::new(num_bytes).with_bitset(num_bytes, false);
+    // determine null buffer, which is a function of `values` and `indices`
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/take.rs:310]: 5 for the rest of the list. We don't benchmark lists, but I expect performance to improve here too

**Comment by jorgecarleitao** [comment]
Sorry, for some reason I did not see your PR here. I will incorporate them, 1s.

**Pull request closed without merging**